lib/dockerregistry/inventory.go

```diff
@@ -2547,22 +2570,41 @@ func (m Manifest) Contains(gcrPayload GCRPubSubPayload) bool {
 			if !strings.Contains(gcrPayload.Digest, (string)(image.ImageName)) {
 				continue
 			}
+			r |= FlagPathMatched
 			for digest, tags := range image.Dmap {
 				fqin := ToFQIN(rc.Name, image.ImageName, digest)
-				if fqinOnly {
+				if len(gcrPayload.Tag) == 0 {
+					// The gcrPayload.Digest field actually holds the FQIN.
 					if gcrPayload.Digest == fqin {
-						return true
+						r |= FlagDigestMatched
 					}
 				} else {
+					// If gcrPayload.Tag is non-empty, then that means a
+					// particular PQIN change occurred in GCR. In this case, it
+					// MUST match both the digest and tag combination as set out
+					// in the manifest.
+					//
+					// Matching solely on the tag (but not digest) or vice versa
+					// goes AGAINST the intent of the promoter manifest and is
+					// cause for alarm!
 					for _, tag := range tags {
 						pqin := ToPQIN(rc.Name, image.ImageName, tag)
+						// The gcrPayload.Tag field actually holds the PQIN.
 						if gcrPayload.Tag == pqin {
-							return gcrPayload.Digest == fqin
+							if gcrPayload.Digest == fqin {
+								r |= FlagDigestMatched
+								r |= FlagTagMatched
+							}
 						}
 					}
 				}
 			}
+			// If we have just a path match, return early because we should
+			// limit the scope of the search to just 1 image.
+			if r&FlagPathMatched != 0 {
+				return r
+			}
 		}
 	}
-	return false
+	return r
 }
```

Comment:
this method is VERY complex, do you think we can maybe simplify it somehow? Maybe spliting it into smaller functions would help?

Reply:
I can split it up at least 1 more level. ACK